Adjust imports in scripts and webpack directories

alanorth opened the pull request:
# Related
- #4632

# Description
A handful of import- and console.log-related fixes were missed in the `scripts` and `webpack` directories due to the lint configuration in `angular.json` not considering them.

## Instructions for Reviewers
Please add a more detailed description of the changes made by your PR. At a minimum, providing a bulleted list of changes in your PR is helpful to reviewers.

List of changes in this PR:
* First, update lint configuration in `angular.json` to consider `scripts` and `webpack` directories
* Second, run `npm run lint-fix` to automatically fix as many errors as possible
* Third, add `node:` prefix to several imports of internal Node.js APIs (fs and path) and scope to methods we use
* Fourth, disable eslint for copy-webpack-plugin because it doesn't support ES Module imports
* Fifth, remove unused sass and lodash imports
* Sixth, rework JSON5, commander, and cli-progress imports to use scoped ES Module imports
* Seventh, use scoped imports for lodash modules
* Eighth, replace `console.log()` usage with `console.info()`

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes.

Make sure the dev and prod builds are successful.

Make sure i18n scripts work:

```
$ npm run merge-i18n -- -s src/themes/custom/assets/i18n
$ npm run sync-i18n
```

Make sure all tests in CI pass.

# Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You do not need to complete this checklist prior creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC9x/Accessibility)** if it makes changes to the user interface.
- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

alexandrevryghem reviewed (changes requested):
Thnx @alanorth for this PR! I already took a quick look at the changed files and it looks great. I did add one small inline suggestion, but I will do a full review once those conflicts are resolved

alanorth committed:
angular.json: lint scripts and webpack dirs

Include the scripts and webpack directories in linting.

alanorth committed:
angular.json: lint scripts and webpack dirs

Include the scripts and webpack directories in linting.

alanorth committed:
Run npm lint-fix

Automatically fix most lint errors.

alanorth committed:
Run npm lint-fix

Automatically fix most lint errors.

alanorth commented:
The scope of this PR has changed because `main` updated the linting rules. Now there are more things to fix in the `scripts` and `webpack` directories. I am working on it.

alanorth committed:
Fix node imports in scripts and webpack dirs

Don't use require-style imports and scope some imports of path and
fs so we only import methods we are using.

alanorth committed:
Fix node imports in scripts and webpack dirs

Don't use require-style imports and scope some imports of path and
fs so we only import methods we are using.

alanorth committed:
webpack: disable eslint for copy-webpack-plugin

Disable the no-require-imports rule for copy-webpack-plugin because
it does not support ES Module style imports.

alanorth committed:
webpack: disable eslint for copy-webpack-plugin

Disable the no-require-imports rule for copy-webpack-plugin because
it does not support ES Module style imports.

alanorth committed:
webpack/webpack.common.ts: remove unused sass import

alanorth committed:
webpack/webpack.common.ts: remove unused sass import

alanorth committed:
Rework json5 import in scripts and webpack

Don't use a require-style import, and only import the methods we
are actually using.

alanorth committed:
Rework json5 import in scripts and webpack

Don't use a require-style import, and only import the methods we
are actually using.

alanorth committed:
scripts: adjust commander imports

Don't use require-style imports, and only import the methods that
we actually use instead of the whole module. Also, we need to use
a different syntax to access the parsed opts after this for some
reason.

alanorth committed:
scripts: adjust commander imports

Don't use require-style imports, and only import the methods that
we actually use instead of the whole module. Also, we need to use
a different syntax to access the parsed opts after this for some
reason.

alanorth committed:
scripts/sync-i18n-files.ts: rework lodash import

Don't use require-style imports, and scope the imports to only the
methods we are using.

alanorth committed:
scripts/sync-i18n-files.ts: rework lodash import

Don't use require-style imports, and scope the imports to only the
methods we are using.

alanorth committed:
scripts/merge-i18n-files.ts: remove unused lodash import

alanorth committed:
scripts/merge-i18n-files.ts: remove unused lodash import

alanorth committed:
scripts: use console.info() instead of console.log()

alanorth committed:
scripts: use console.info() instead of console.log()

alanorth committed:
scripts: adjust import of cli-progress

Don't use require-style imports, and only import the methods that
we are actually using.

alanorth committed:
scripts: adjust import of cli-progress

Don't use require-style imports, and only import the methods that
we are actually using.

alanorth committed:
scripts/env-to-yaml.ts: disable eslint rule

Disable eslint rule for require-style imports here since this isn't
importing a Node.js module, but a command line argument.

alanorth committed:
scripts/env-to-yaml.ts: disable eslint rule

Disable eslint rule for require-style imports here since this isn't
importing a Node.js module, but a command line argument.

alanorth commented:
I'm not sure why the lodash import rule is triggering, since we are importing individual methods:

```ts
import {
  isEmpty,
  isEqual,
} from 'lodash';
```

Must be something subtle about the plugin's configuration? @alexandrevryghem?

alexandrevryghem commented:
@alanorth: It's part of the `lodash/import-scope` rule. The correct usage is:

```ts
import isEmpty from 'lodash/isEmpty';
import isEqual from 'lodash/isEqual';
```

alanorth commented:
Thanks @alexandrevryghem. I've updated it.

For what it's worth, I see our configuration of `lodash/import-scope` uses the "method" import style:

```ts
import isEmpty from 'lodash/isEmpty';
import isEqual from 'lodash/isEqual';
```

While this style would be the "member" style:

```ts
import {
  isEmpty,
  isEqual,
} from 'lodash';
```

All of our other imports in dspace-angular use the latter style.

P.S. removing the `port to dspace-9_x` tag since this ended up being more changes than I thought (and for example, the `console.log()` linting changes were not ported there).

alanorth commented:
Something strange going on with tests in CI. Closing and re-opening.

alanorth commented:
I'm not sure why this is failing e2e tests in CI. I see some error early in the logs after the e2e tests start:

```console
2026-04-22T07:09:03.5171951Z 
2026-04-22T07:09:03.5287942Z Environment extended with app config
2026-04-22T07:09:03.5289193Z 
2026-04-22T07:09:03.5290717Z dspace-angular
2026-04-22T07:09:03.5293112Z Version: 10.0.0-next
2026-04-22T07:09:03.5293526Z Environment: Production
2026-04-22T07:09:03.5293772Z 
2026-04-22T07:09:03.5319495Z [HPM] Proxy created: /  -> http://127.0.0.1:8080/server/sitemaps
2026-04-22T07:09:03.5323004Z [HPM] Proxy created: /  -> http://127.0.0.1:8080/server
2026-04-22T07:09:03.5629110Z [07:09:03 GMT+0000 (Coordinated Universal Time)] Listening at http://127.0.0.1:4000
2026-04-22T07:09:04.2817375Z ERROR SyntaxError: Expected property name or '}' in JSON at position 1
2026-04-22T07:09:04.2817879Z     at JSON.parse (<anonymous>)
2026-04-22T07:09:04.2818547Z     at TranslateServerLoader.getTranslation (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:7801704)
2026-04-22T07:09:04.2819540Z     at TranslateService2.loadAndCompileTranslations (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:5256381)
2026-04-22T07:09:04.2820699Z     at TranslateService2.loadOrExtendLanguage (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:5256092)
2026-04-22T07:09:04.2821932Z     at TranslateService2.use (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:5255608)
2026-04-22T07:09:04.2822681Z     at Object.next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:591711)
2026-04-22T07:09:04.2823423Z     at ConsumerObserver2.next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3336710)
2026-04-22T07:09:04.2824246Z     at SafeSubscriber2.Subscriber2._next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3336072)
2026-04-22T07:09:04.2825101Z     at SafeSubscriber2.Subscriber2.next (/home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3335509)
2026-04-22T07:09:04.2825813Z     at /home/runner/work/dspace-angular/dspace-angular/dist/server/main.js:1:3419624
```

This is followed by a bunch of other errors about reading undefined from a pipe. I haven't figured out why this is happening yet. I suppose it is all due to the first JSON parsing error, seemingly in the translation service?

MMilosz commented on webpack/webpack.common.ts:
```diff
@@ -1,41 +1,51 @@
       },
       transform(content) {
-        return JSON.stringify(JSON5.parse(content.toString()));
-      }
+        return stringify(parse(content.toString()));
+      },
```
This change makes source `JSON5` files output as `JSON5` instead of `JSON`, which makes **syntax errors inside i18n JSON files** (interestingly, it was caught in only 5 out of 50+ e2e tests)

That's exactly why you get errors like `ERROR SyntaxError: Expected property name or '}' in JSON at position 1`

I changed it locally to `return JSON.stringify(parse(content.toString())` and it fixed all e2e tests

alanorth replied:
Thanks @MMilosz! Actually, this was my mistake when modifying the `JSON5` imports to only import the modules we use. I mistakenly modified the `JSON` one! I have verified it works locally and amended my commits, but not tried rebasing for latest `main` (DSpace 11-SNAPSHOT).

BTW I think your issue with the i18n strings was due to an outdated dist dir? I ran `npm run clean:dist` before building and I see the strings for all languages.

MMilosz replied:
Hm, no idea. I usually clone fresh repo for PRs to minimize such nuances. CI also had i18n issues before (logs reported error with `}`)

Anyways! Checked it now, all looks perfectly fine in both dev and prod envs!

MMilosz reviewed (changes requested):
Hi, I was reading article about [replacing lodash with lodash-es](https://dev.to/pffigueiredo/making-lodash-tree-shakable-3h27) and found this PR in a result. The failing e2e tests looked suspicious until I ran them locally, seeing i18n keys missing:

<img width="1065" height="161" alt="Image" src="https://github.com/user-attachments/assets/7b20658c-20d4-4e69-9f5e-9274b6d5a55b" />

I ctrl+f 'json' in the PR & found change in webpack.common.ts `transform(content)`. @alanorth after applying the fix (see comment), all e2e tests passed locally for me

EDIT: I rebased your changes from this PR into the latest main and confirm it generally works! (except the i18n issue)

MMilosz commented on angular.json:
```diff
@@ -271,7 +271,9 @@
-              "src/**/*.json5"
+              "src/**/*.json5",
+              "scripts/*.ts",
+              "webpack/*.ts"
```
Is `/*.ts` (instead of `/**/*.ts`) intentional? In this shape, it misses `*.ts` files in e.g. `webpack/plugins/` and `scripts/config/` (where we already have some ts files)

alanorth replied:
@MMilosz good catch! I updated the pattern and applied the fixes to my branch. Thanks.

MMilosz reviewed (approved):
LGTM! Always great to see work being done to modernize the frontend tooling

I left one minor detail regarding the glob pattern in angular.json, but overall this is a great cleanup

alanorth committed:
webpack/plugins: auto fix one more import

tdonohue reviewed (approved):
👍 Thanks @alanorth !  This looks great to me too.  I tested it by redeploying the UI, and found no issues in SSR.  Also tested the various commandline tools like `npm run merge-i18n`, `npm run sync-i18n` and `npm run test:rest`.  I didn't see any changes in behavior.
